types/decimal.go

```diff
@@ -1,138 +1,150 @@
+		fracPart = i % int64(math.Pow10(-exponent)) * int64(math.Pow10(4+exponent))
+	} else {
+		intPart = i * int64(math.Pow10(exponent))
+		if i > 0 && intPart < i {
+			return Decimal{}, fmt.Errorf("%w: value %ve%v would overflow", ErrDecimal, i, exponent)
+		} else if i < 0 && intPart > i {
+			return Decimal{}, fmt.Errorf("%w: value %ve%v would underflow", ErrDecimal, i, exponent)
 		}
 	}
 
-	// Parse the required first digit.
-	c := rune(s[i])
-	if !unicode.IsDigit(c) {
-		return Decimal{}, fmt.Errorf("%w: unexpected character %s", ErrDecimal, strconv.QuoteRune(c))
-	}
-	integer := int64(c - '0')
-	i++
+	return newDecimal(intPart, int16(fracPart))
+}
 
-	// Parse any other digits, ending with i pointing to '.'.
-	for ; ; i++ {
-		if i == len(s) {
-			return Decimal{}, fmt.Errorf("%w: string missing decimal point", ErrDecimal)
-		}
-		c = rune(s[i])
-		if c == '.' {
-			break
-		}
-		if !unicode.IsDigit(c) {
-			return Decimal{}, fmt.Errorf("%w: unexpected character %s", ErrDecimal, strconv.QuoteRune(c))
-		}
-		integer = 10*integer + int64(c-'0')
-		if integer > 922337203685477 {
-			return Decimal{}, fmt.Errorf("%w: overflow", ErrDecimal)
-		}
+// NewDecimalFromInt returns a Decimal with the whole integer value provided
+func NewDecimalFromInt[T constraints.Signed](i T) (Decimal, error) {
+	return NewDecimal(int64(i), 0)
+}
+
+// NewDecimalFromFloat returns a Decimal that approximates the given floating point value.
+// The value of the Decimal is calculated by multiplying it by 10^4, truncating it to
+// an int64 representation to cut off any digits beyond the four allowed, and passing it
+// as an integer to NewDecimal() with -4 as the exponent.
+//
+// WARNING: decimal representations of more than 6 significant digits for float32s and 15
+// significant digits for float64s can be lossy in terms of precision. To create a precise
+// Decimal above those sizes, use the NewDecimal constructor.
+func NewDecimalFromFloat[T constraints.Float](f T) (Decimal, error) {
+	f = f * decimalPrecision
+	if f > math.MaxInt64 {
+		return Decimal{}, fmt.Errorf("%w: value %v would overflow", ErrDecimal, f)
+	} else if f < math.MinInt64 {
+		return Decimal{}, fmt.Errorf("%w: value %v would underflow", ErrDecimal, f)
 	}
 
-	// Advance past the '.'.
-	i++
+	return Decimal{int64(f)}, nil
+}
+
+// Compare returns
+//
+//	-1 if d is less than other,
+//	 0 if d equals other,
+//	+1 if d is greater than other.
+func (d Decimal) Compare(other Decimal) int {
+	return cmp.Compare(d.value, other.value)
+}
 
-	// Parse the fraction part
-	fraction := int64(0)
-	fractionDigits := 0
-	for ; i < len(s); i++ {
-		c = rune(s[i])
-		if !unicode.IsDigit(c) {
-			return Decimal{}, fmt.Errorf("%w: unexpected character %s", ErrDecimal, strconv.QuoteRune(c))
+// ParseDecimal takes a string representation of a decimal number and converts it into a Decimal type.
+func ParseDecimal(s string) (Decimal, error) {
+	decimalIndex := strings.Index(s, ".")
+	if decimalIndex < 0 {
```

Comment:
Are we sure the "." is required?

Reply:
Good question. The Cedar Rust implementation [does seem to require it](https://github.com/cedar-policy/cedar/blob/dc9163c6f71d72633495f80262ab2d0b266138f7/cedar-policy-core/src/extensions/decimal.rs#L59).